sdk/src/main/scala/io/horizen/account/sc2sc/AbstractCrossChainRedeemMessageProcessor.scala

```diff
@@ -0,0 +1,135 @@
+package io.horizen.account.sc2sc
+
+import com.google.common.primitives.Bytes
+import io.horizen.account.state.events.AddCrossChainRedeemMessage
+import io.horizen.account.state.{BaseAccountStateView, ExecutionRevertedException, Message, NativeSmartContractMsgProcessor}
+import io.horizen.cryptolibprovider.{CryptoLibProvider, Sc2scCircuit}
+import io.horizen.params.NetworkParams
+import io.horizen.sc2sc.{CrossChainMessage, CrossChainMessageHash}
+import io.horizen.utils.BytesUtils
+import sparkz.crypto.hash.Keccak256
+
+trait CrossChainRedeemMessageProvider {
+  def doesCrossChainMessageHashFromRedeemMessageExist(hash: CrossChainMessageHash, view: BaseAccountStateView): Boolean
+}
+
+abstract class AbstractCrossChainRedeemMessageProcessor(
+                                                         networkParams: NetworkParams,
+                                                         sc2scCircuit: Sc2scCircuit,
+                                                         scTxMsgProc: ScTxCommitmentTreeRootHashMessageProvider
+                                                       ) extends NativeSmartContractMsgProcessor with CrossChainRedeemMessageProvider {
+  protected def processRedeemMessage(msg: AccountCrossChainRedeemMessage, view: BaseAccountStateView): Array[Byte] = {
+    validateRedeemMsg(msg, view)
+
+    addCrossChainMessageToView(view, msg)
+
+    addCrossChainRedeemMessageLogEvent(view, msg)
+
+    msg.encode()
+  }
+
+  private def addCrossChainRedeemMessageLogEvent(view: BaseAccountStateView, accCcRedeemMessage: AccountCrossChainRedeemMessage): Unit = {
+    val event = AddCrossChainRedeemMessage(
+      accCcRedeemMessage.accountCrossChainMessage,
+      accCcRedeemMessage.certificateDataHash,
+      accCcRedeemMessage.nextCertificateDataHash,
+      accCcRedeemMessage.scCommitmentTreeRoot,
+      accCcRedeemMessage.nextScCommitmentTreeRoot,
+      accCcRedeemMessage.proof
+    )
+    val evmLog = getEthereumConsensusDataLog(event)
+    view.addLog(evmLog)
+  }
+
+  private def addCrossChainMessageToView(view: BaseAccountStateView, accCcRedeemMessage: AccountCrossChainRedeemMessage): Unit = {
+    val messageHash = CryptoLibProvider.sc2scCircuitFunctions.getCrossChainMessageHash(
+      AbstractCrossChainMessageProcessor.buildCrosschainMessageFromAccount(accCcRedeemMessage.accountCrossChainMessage, networkParams)
+    )
+    setCrossChainMessageHash(messageHash, view)
+  }
+
+  /**
+   * Extract the AccountCrossChainRedeemMessage from the message
+   *
+   * @param msg message to apply to the state
+   * @return the AccountCrossChainRedeemMessage built from the arguments in the message
+   */
+  protected def getAccountCrossChainRedeemMessageFromMessage(msg: Message): AccountCrossChainRedeemMessage
+
+  private def validateRedeemMsg(ccRedeemMgs: AccountCrossChainRedeemMessage, view: BaseAccountStateView): Unit = {
+    try {
```

Comment:
Does it make sense to check also `messageType` value of `accountCcMsg`?

Moreover, I understand that `sender/receiver` have a generic byte array format therefore are not checked on purpose, correct?

But however consider checking them not to be null or zero-length.
Finally, does it make sense for them to have a maximum size in order to prevent malicious behaviour?

Reply:
We don't enforce any check on `messageType` at this level, this is something that the implementation will do.
Correct about the `sender/receiver`, at this level we're just interested into checking the receiving sidechain is correct; however we could check whether they're null, empty or too large.